python/tvm/relay/grammar/Relay.g4

```diff
@@ -74,106 +73,121 @@ METADATA: 'METADATA:' .*;
+  | '{' expr '}'                                 # paren
   // function application
-  | expr '(' callList ')'                     # call
-  | '-' expr                                  # neg
-  | expr op=('*'|'/') expr                    # binOp
-  | expr op=('+'|'-') expr                    # binOp
-  | expr op=('<'|'>'|'<='|'>=') expr          # binOp
-  | expr op=('=='|'!=') expr                  # binOp
+  | expr '(' callList ')'                        # call
+  | '-' expr                                     # neg
+  | expr op=('*'|'/') expr                       # binOp
+  | expr op=('+'|'-') expr                       # binOp
+  | expr op=('<'|'>'|'<='|'>=') expr             # binOp
+  | expr op=('=='|'!=') expr                     # binOp
   // function definition
-  | func                                      # funcExpr
+  | func                                         # funcExpr
   // tuples and tensors
-  | '(' ')'                                   # tuple
-  | '(' expr ',' ')'                          # tuple
-  | '(' expr (',' expr)+ ')'                  # tuple
-  | expr '.' NAT                              # projection
-  | '[' (expr (',' expr)*)? ']'               # tensor
-  | 'if' '(' expr ')' body 'else' body        # ifElse
+  | '(' ')'                                      # tuple
+  | '(' expr ',' ')'                             # tuple
+  | '(' expr (',' expr)+ ')'                     # tuple
+  | '[' (expr (',' expr)*)? ']'                  # tensor
+  | 'if' '(' expr ')' body 'else' body           # ifElse
+  | matchType '(' expr ')' '{' matchClause+ '}'  # match
+  | expr '.' NAT                                 # projection
   // sequencing
-  | 'let' var '=' expr ';' expr               # let
+  | 'let' var '=' expr ';' expr                  # let
   // sugar for let %_ = expr; expr
-  | expr ';;' expr                            # let
-  | GRAPH_VAR '=' expr ';' expr               # graph
-  | ident                                     # identExpr
-  | scalar                                    # scalarExpr
-  | meta                                      # metaExpr
-  | QUOTED_STRING                             # stringExpr
+  | expr ';;' expr                               # let
+  | graphVar '=' expr ';' expr                   # graph
+  | ident                                        # identExpr
+  | scalar                                       # scalarExpr
+  | meta                                         # metaExpr
+  | QUOTED_STRING                                # stringExpr
+  ;
+
+func: 'fn' typeParamList? '(' argList ')' ('->' typeExpr)? body ;
+defn
+  : 'def' globalVar typeParamList? '(' argList ')' ('->' typeExpr)? body  # funcDefn
+  | 'type' generalIdent typeParamList? '=' adtConsDefn+                      # adtDefn
+  ;
+
+constructorName: CNAME ;
+
+adtConsDefn: '|' constructorName ('(' typeExpr (',' typeExpr)* ')')? ;
+matchClause: '|' constructorName patternList? '=>' expr ;
+matchType : 'match' | 'match?' ;
```

Comment:
I think you should clearly document the difference between these matches (complete vs incomplete) and ensure it round-trips through the pretty-printer

Reply:
Ah, I notice the pretty-printer does handle it :+1: